### PR TITLE
Fixed conflicting styles for product image container in different listings at the same page

### LINKED ISSUE #31526: [Issue] Fixed conflicting styles for product image container in different listings at the same page
This issue is automatically created based on existing pull request: magento/magento2#31473: Fixed conflicting styles for product image container in different listings at the same page
## 

### Description (*)

This pull requests solves problem when there are two product listing blocks on the same page. Both blocks have at list one common product. And listings have different image sizes for product image.

### Related Pull Requests

Can't find related pull requests.

### Fixed Issues (if relevant)

Can't find related issue.

### Manual testing scenarios (*)

1. Open product "Crown Summit Backpack" (SKU 24-MB03) to assign up-sell products. Add at list one up-sell product. One of the assigned products must be "Fusion Backpack" (SKU 24-MB02). Save changes.
1. Create widget "Catalog Product List" via Magento Admin. Assign widget to store view you test. Layout updates settings: display on "Specified page"; page "Catalog Product View"; container "Main Content Top". Widget Options - select at least two product via SKU. One of products must be "Fusion Backpack" (SKU 24-MB02). Save widget.
1. Clear Magento 2 cache.
1. Open product page for "Crown Summit Backpack" (SKU 24-MB03) at store front.

image

With old code you can see that "Fusion Backpack" has smaller image than it should.

This pull request fixes it.  "Fusion Backpack" will get same size as "Affirm Water Bottle".

### Contribution checklist (*)
 - < ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ > All automated tests passed successfully (all builds are green)

### LINKED ISSUE #35372: [Issue] Fixed conflicting styles for product image container in different listings at the same page
This issue is automatically created based on existing pull request: magento/magento2#31473: Fixed conflicting styles for product image container in different listings at the same page

---------

### Description (*)

This pull requests solves problem when there are two product listing blocks on the same page. Both blocks have at list one common product. And listings have different image sizes for product image.

### Related Pull Requests

Can't find related pull requests.

### Fixed Issues (if relevant)

Can't find related issue.

### Manual testing scenarios (*)

1. Open product "Crown Summit Backpack" (SKU 24-MB03) to assign up-sell products. Add at list one up-sell product. One of the assigned products must be "Fusion Backpack" (SKU 24-MB02). Save changes.
2. Create widget "Catalog Product List" via Magento Admin. Assign widget to store view you test. Layout updates settings: display on "Specified page"; page "Catalog Product View"; container "Main Content Top". Widget Options - select at least two product via SKU. One of products must be "Fusion Backpack" (SKU 24-MB02). Save widget.
3. Clear Magento 2 cache.
4. Open product page for "Crown Summit Backpack" (SKU 24-MB03) at store front.

image

With old code you can see that "Fusion Backpack" has smaller image than it should.

This pull request fixes it.  "Fusion Backpack" will get same size as "Affirm Water Bottle".

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#31526: Fixed conflicting styles for product image container in different listings at the same page

### PR DESCRIPTION
**Pull request by tangar76**
### Description (*)

This pull requests solves problem when there are two product listing blocks on the same page. Both blocks have at list one common product. And listings have different image sizes for product image.

### Related Pull Requests

Can't find related pull requests.

### Fixed Issues (if relevant)

Can't find related issue.

### Manual testing scenarios (*)

1. Open product "Crown Summit Backpack" (SKU 24-MB03) to assign up-sell products. Add at list one up-sell product. One of the assigned products must be "Fusion Backpack" (SKU 24-MB02). Save changes.
2. Create widget "Catalog Product List" via Magento Admin. Assign widget to store view you test. Layout updates settings: display on "Specified page"; page "Catalog Product View"; container "Main Content Top". Widget Options - select at least two product via SKU. One of products must be "Fusion Backpack" (SKU 24-MB02). Save widget.
3. Clear Magento 2 cache.
4. Open product page for "Crown Summit Backpack" (SKU 24-MB03) at store front.

image

With old code you can see that "Fusion Backpack" has smaller image than it should.

This pull request fixes it.  "Fusion Backpack" will get same size as "Affirm Water Bottle".

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#31526: Fixed conflicting styles for product image container in different listings at the same page

### Resolved issues:
1. [x] resolves magento/magento2#35372: Fixed conflicting styles for product image container in different listings at the same page

### CODE REVIEW

**Comment by sidolov** [comment]
@magento create issue

**Comment by gabrieldagama** [comment]
@magento run all tests

File: app/code/Magento/Catalog/view/frontend/templates/product/image_with_borders.phtml
Comment (gabrieldagama) [inline app/code/Magento/Catalog/view/frontend/templates/product/image_with_borders.phtml]: I believe we should create a function on `Magento\Catalog\Block\Product\Image` to generate the containerId instead of adding it to template.

**Review by gabrieldagama — CHANGES_REQUESTED** [review]
Hi @tangar76, thanks for your contribution! Can you have a look at my comment below?

Also, we must cover all changes with automated tests, probably the most suitable here would be MFTF test.

Thanks!

**Comment by engcom-Charlie** [comment]
Hello @tangar76, Could you please look at https://github.com/magento/magento2/pull/31473#discussion_r551856974?
Thank you.

**Comment by engcom-Charlie** [comment]
Hello @tangar76, please look at https://github.com/magento/magento2/pull/31473#discussion_r551856974. Otherwise, we can't proceed with your PR.
Thank you.

**Comment by tangar76** [comment]
@engcom-Charlie container id logic moved from template to block

**Comment by gabrieldagama** [comment]
@magento run all tests

File: app/code/Magento/Catalog/Block/Product/Image.php
Comment (gabrieldagama) [inline app/code/Magento/Catalog/Block/Product/Image.php]: Please add the return type string to the function signature.

**Comment by gabrieldagama** [comment]
@magento run all tests

**Review by gabrieldagama — CHANGES_REQUESTED** [review]
Hi @tangar76, thanks for moving the logic to the block, can you have a look at my comment below? Also, can you cover these changes with automated tests? MFTF may be suitable here!

Thanks!

**Comment by tangar76** [comment]
@gabrieldagama , I've added type return type for method getContainerId in product image renderer block.

Unfortunately this pull request has to be postponed. I can't cover it with MFTF test anytime soon. I will return to it in near future.

**Comment by gabrieldagama** [comment]
@tangar76 No worries, thanks for contributing!

@engcom-Charlie can you help handle the test coverage on this one?

Thanks!

**Comment by engcom-Charlie** [comment]
@tangar76 @gabrieldagama  I'll take care of the test coverage.

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Review by gabrieldagama — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @gabrieldagama, thank you for the review.
[ENGCOM-8786](https://jira.corp.magento.com/browse/ENGCOM-8786) has been created to process this Pull Request
:eight_spoked_asterisk: @gabrieldagama, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by magento-engcom-team** [comment]
@tangar76 thank you for contributing. Please accept Community Contributors team invitation [here](https://github.com/orgs/magento/invitation) to gain extended permissions for this repository.

**Comment by engcom-Bravo** [comment]
:heavy_check_mark: QA PAssed

### Manual Testing Scenario
1. Install Sample Data
2. From Admin Panel, Go to _Catalog - Products_ and open product "Crown Summit Backpack" (SKU 24-MB03) 
3. Expand **Related Products, Up-Sells, and Cross-Sells** and add several _Up-Sell Products_ using **Add Up-Sell Products**.
One of the assigned products must be "Fusion Backpack" (SKU 24-MB02).
crown_uppsell
4. Go to _Content - Widgets_ and click **Add Widget**
5. Set the following settings:
- Catalog Product List 
- Assign widget to store view 
- Layout updates settings: 
     - display on "Specified page"; 
     - page "Catalog Product View"; 
     - container "Main Content Top".
5. Switch to _Widget Options_ and set **Condition** _SKU  **is one of**  24-MB02,24-UG06_
widget_condition
6. **Save** Widget, perform `bin/magento indexer:reindex; bin/magento cache:clean`
7. Go to Storefront and open the "Crown Summit Backpack" (SKU 24-MB03) products page

**BEFORE** applying changes provided in the PR, the product images had different sizes 
before

**AFTER** switching to the PR, the product images have the same sizes
after

Added more than two products
multiple

**Comment by engcom-Charlie** [comment]
Moved in `To Approve` since **SVC** falls with error:
the public method has been added to `Magento\Catalog\Block\Product\Image`.

**Comment by template-provider** [comment]
+1
I would like to see this merged ... what is the problem?

**Comment by engcom-Alfa** [comment]
@magento run all tests

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Preconditions:**

1. Have Magento latest code installed with Sample data

**Manual testing scenario:**

1. Open product "Crown Summit Backpack" (SKU 24-MB03) to assign up-sell products. Add at list one up-sell product. One of the assigned products must be "Fusion Backpack" (SKU 24-MB02). Save changes.
2. Create widget "Catalog Product List" via Magento Admin. Assign widget to store view you test. Layout updates settings: display on "Specified page"; page "Catalog Product View"; container "Main Content Top". Widget Options - Set the Condition to Weight-1. Save widget.
3. Open the Fusion Backpack's edit product page, set the product weight to 1 and save. Also pickup any other 1/2 more products and set the weight to 1.
4. Clear Magento 2 cache.
5. Open product page for "Crown Summit Backpack" at store front.

**Before:** :heavy_multiplication_x: One line description about what was happening prior
image

**After:** :heavy_check_mark:  One line description about the is the outcome after fix
image

Additionally tested by adding a couple of more products to fall under the Widget list.
Also noticed another product also affected with image size and corrected after the fix. Below are the screenshots with more products.

Before the PR fix:
image

After the PR fix:
image

**Comment by engcom-Alfa** [comment]
@magento run Functional Tests B2B, Functional Tests EE, WebAPI Tests, Semantic Version Checker

**Comment by engcom-Charlie** [comment]
Raised internal approval request JIRA [ticket](https://jira.corp.magento.com/browse/MC-43361) for MVC test failure. Waiting for the approvals.

**Comment by engcom-Alfa** [comment]
@magento run all tests

**Comment by engcom-Alfa** [comment]
As we have got code updates later the QA validations, we have to re-test the scenario once. Re-considered for testing.
Will update here shortly!

**Comment by engcom-Alfa** [comment]
@magento run Functional Tests B2B, Functional Tests EE, Semantic Version Checker, Static Tests, WebAPI Tests

**Comment by engcom-Alfa** [comment]
@magento run Functional Tests B2B, Functional Tests EE

**Comment by engcom-Alfa** [comment]
@magento run Static Tests

**Comment by engcom-Alfa** [comment]
Re-retested again, and QA passed where the report remains same as mentioned in [above-comment](https://github.com/magento/magento2/pull/31473#issuecomment-1106286872).
Due to automation job failures, moving to _Extended Testing_.

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests B2B, Functional Tests EE, Semantic Version Checker, Static Tests

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests B2B, Functional Tests EE, Semantic Version Checker, Static Tests

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests B2B, Functional Tests EE, Semantic Version Checker

**Comment by engcom-Charlie** [comment]
In order to proceed with this PR, merged the latest code and run the build.

Gone through the build failures.

**For functional test failures:**  The tests which are failing in build are not part of this PR or not failing because of any change.  Also all the failing functional tests are working fine on local PR environment. 
image


**For Semantic Version Checker**: As mentioned in above comment for SVC failure raised approval jira ticket. 

Until we get the approval, moving this PR in Pending Approval.

Thank you!

**Comment by engcom-Charlie** [comment]
Raised new internal approval request JIRA [ticket](https://jira.corp.adobe.com/browse/AC-6876) for MVC test failure. Waiting for the approvals.

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
looks like this stale does not get enough attention

**Comment by engcom-Charlie** [comment]
The remarks we got today in the approval [jira](https://jira.corp.adobe.com/browse/AC-6876) is: "The initial issue can be resolved by adding ViewModel instead of the new method. Canceling the request, the solution should be re-implemented."

As per above the approval jira is cancelled and this PR needs reimplementation. Hence moving it to Change Requested status.

Thank you!

**Comment by engcom-Charlie** [comment]
To take care the Semantic version checker failure, I am  working on the new implementation with ViewModel as mentioned in above [comment](https://github.com/magento/magento2/pull/31473#issuecomment-1447624720).

Thank you!

**Comment by engcom-Lima** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
It seems that bot has moved this PR to Ready for testing hence moving it back to Changes Requested.

Thank you!

**Comment by engcom-Charlie** [comment]
It seems that bot has moved this PR to Ready for testing hence moving it back to Changes Requested.

Thank you!

**Comment by engcom-Charlie** [comment]
Implemented the fix for the issue using View model as request in internal approval [jira](https://github.com/magento/magento2/pull/31473#issuecomment-1447624720).

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests CE

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests B2B, Functional Tests EE

**Comment by engcom-Charlie** [comment]
@magento give me test instance

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento give me 2.4-develop instance

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests CE

**Comment by engcom-Charlie** [comment]
Hi @tangar76,

There are many failures in the build, it will be helpful if you have a look into them. Can you please check them.

Thank you!

**Comment by engcom-Charlie** [comment]
Hi @tangar76,

As per above [comment](https://github.com/magento/magento2/pull/31473#issuecomment-1719481216), there are too many failures in this PR. Kindly have a look into this. Till then we are closing this PR. 

Please feel free to reopen once it get updated or let us know to reopen when you are working on it.

Thank you!

**Pull request closed without merging**
